crypto/multisig/src/lib.rs

```diff
@@ -69,30 +69,62 @@ impl<
     }
 }
 
-/// A set of M-out-of-N public keys.
+/// A set of M-out-of-N signer identities, where a signer identity can be either
+/// a public key or a nested M-out-of-N set of identities.
 #[derive(
     Clone, Deserialize, Digestible, Eq, Hash, Message, Ord, PartialEq, PartialOrd, Serialize,
 )]
 #[serde(bound = "")]
 pub struct SignerSet<P: Default + PublicKey + Message> {
-    /// List of potential signers.
+    /// List of potential individual signers.
     #[prost(message, repeated, tag = "1")]
-    signers: Vec<P>,
-
-    /// Minimum number of signers required.
+    #[digestible(name = "signers")]
+    individual_signers: Vec<P>,
+
+    /// List of potential sets of signers. This allows us to declare a nested
+    /// multi-signature scheme. Ideally the two arrays in this struct would be
+    /// combined into a single one that holds an enum of (IndividualSigner,
+    /// MultiSignerSet), but since this feature was added after we went live
+    /// and instances of SignerSet made their way into the ledger, it is not
+    /// possible to change the array type without breaking backwards
+    /// compatibility. This is also the reason the tag numbers in the struct
+    /// are not sequential.
+    #[prost(message, repeated, tag = "3")]
+    multi_signers: Vec<SignerSet<P>>,
+
+    /// Minimum number of signers required. The potential signers are the union
+    /// of `sigenrs` and `signer_sets` This implies that the upper limit
+    /// (the total number of possible signers) is `signers.len() +
+    /// signer_sets.len()`).
```

Comment:
Old code did not have multi signers so I am not sure how to answer this question. Am I misunderstanding it?

Reply:
"old code" == code currently out in the wild.
I'm assuming it knows about the single signer list and the threshold.   Getting a new protobuf the old code will only pull in the single signer list and the threshold.

Reply:
Ah, I see. Yes, that observation is correct. I don't expect it to be a problem. `SignerSet`s that contain a multisig can show up in one of two places:
1) Inside consensus-service - we upgrade the entire consensus network at once, and all nodes have to agree with eachother about the hash of the governors (amongst other configuration data). So for consensus nodes it is not a problem since they all get upgraded at once to a version that supports the new field.
2) Inside blocks - blocks store `ValidatedMintConfigTx` and `MintConfig` objects, both containing SignerSets. Old code encountering blocks that make use of the new multi signers field will refuse appending such blocks, since the block contents hash will be incorrectly calculated. That does mean that the launch of this feature has to be coupled with a block version increase (which is a nice way of letting clients know its time to upgrade their code). But that is part of the plan. Part of the work on this feature is to add code that prevents the use of multi-signers in the `tokens.json` file or `MintConfig`s below a certain block version. I am planning on doing that in a follow up PR.

Reply:
I initially read
> we upgrade the entire consensus network at once

and thought then why do we care about backwards compatibility?
After thinking about it, it's so that new code is able to read old blocks.
Thanks for the info!